### ISSUE TITLE
GSE171668, GSE163530 A single-cell and spatial atlas of autopsy tissues reveals pathology and cellular targets of SARS-CoV-2

### ISSUE THREAD

**Issue by rays22** [ebi-ait/hca-ebi-wrangler-central#317]
### Google Sheet:
https://docs.google.com/spreadsheets/d/1DKVGFAwP3bcA4Atk9kW2LQe4rKahWW4B/edit#gid=1189631159

### Project Submission details:
GSE171668_SingleCellAndSpatialAtlasCovid
Project UUID: [61515820-5bb8-45d0-8d12-f0850222ecf0](https://contribute.data.humancellatlas.org/projects/detail?uuid=61515820-5bb8-45d0-8d12-f0850222ecf0)
Submission UUID: [9c4eea83-8939-4232-9568-d31eea06718d](https://contribute.data.humancellatlas.org/submissions/detail?uuid=9c4eea83-8939-4232-9568-d31eea06718d&project=61515820-5bb8-45d0-8d12-f0850222ecf0)
API: [626050ad6357205eb7bbc1e5](https://api.ingest.archive.data.humancellatlas.org/submissionEnvelopes/626050ad6357205eb7bbc1e5)

### Primary Wrangler:
Ami

### Secondary Wrangler:
Wei

### Published study links

* HCA publication:
https://www.nature.com/articles/s41586-021-03570-8
COVID-19 tissue atlases reveal SARS-CoV-2 pathology and cellular targets.

* Accessioned data:
GSE163530,GSE171668,GSE162911,GSE164013
DUOS-000126, DUOS-000127, DUOS-000128, DUOS-000129

### Key Events

- [x] Convert published metadata to HCA spreadsheet
- [x] Manually curate dataset to meet HCA metadata standard
- [x] Collect any matrix and cell-type annotation files
- [x] Upload sheet to validate metadata
- [x] Check linking using ingest graph validator
- [ ] Transfer raw files to ingest to validate data files
- [x] Ask the Secondary Wrangler for an end-to-end review of the project. Ask the Expertise Wrangler to review specific tabs if needed
- [ ] Submit dataset to Production 
- [ ] Convert project data to SCEA format following the SCEA conversion SOP if appropriate

**Comment by ami-day** [comment]
Error when uploading to ingest prod., possibly because of known issues with imaging tabs cause problems with import: https://contribute.data.humancellatlas.org/submissions/detail?uuid=3cf32c24-c2a8-4874-a413-aeb1ab66b2af

**Comment by ami-day** [comment]
Keeping this in the secondary review column as it still needs secondary review while the imaging tabs issues are investigated.

**Comment by Wkt8** [comment]
Assigning self for secondary review.

**Comment by ami-day** [comment]
We can now upload the imaging tabs but I am unable to validate this dataset because the following fields are currently required in the imaging protocol tab:

* should have required property 'numerical_aperture' at root of document
* should have required property 'pixel_size' at root of document
* should have required property 'overlapping_tiles' at root of document 

This information isn't available in the publication. I think these fields should not be required. The sequencing data can be analysed without knowledge of the specific microscopy settings.

**Comment by Wkt8** [comment]
![Screenshot 2022-01-25 at 16.46.01.png](https://images.zenhubusercontent.com/5fd74f67b86eb46030b3f4e3/7e187584-149d-4fb9-b494-797b097f14d9)

**Comment by ami-day** [comment]
The metadata schema update needed to validate this dataset (and other imaging datasets) has been prioritised.

**Comment by prabh-t** [comment]
Wei will work on this.

**Comment by MightyAx** [comment]
@Wkt8 to secondary review

**Comment by Wkt8** [comment]
Waiting for the up-to-date version of the spreadsheet

**Comment by Wkt8** [comment]
Hi! some comments, especially on the linking of the imaging protocols, but overall it looks good! It's a big dataset - and the image_tab information is still missing (and I think there'll be quite a lot of information there!). 
Also, a quick question. I noticed that the cell_suspensions were linked directly to an analysis_file, rather than to sequence_files. I assume the sequence_files were unavailable? 

donor_organism: 
The organisms with age of 80 are actually >=80, so maybe put a range for 80 - 100? 

Specimen_from_organism
There are imaging preparation protocols, and imaging protocols linked to the process generating specimen_from_organism. 

These should be removed as they should be linked to the process relating to the generation of imaged_specimens and image_file, respectively. 

Imaged_specimen
The imaging protocol IDs linked in this tab should be removed, and put in the image_file tab once we have gotten information back on the image_files. 

Imaging_preparation_protocol
For 'Preservation_method' instead of 'FFPE' it should be 'formalin fixed and paraffin embedded', as that is what we have as an enum in our schema 

Imaging_protocol
The 'Microscopy_technique' and 'magnification' needs to be filled in for the other imaging_protocols

Image_file
There is a single row which says 'emailed the authors'?

**Comment by ami-day** [comment]
> Hi! some comments, especially on the linking of the imaging protocols, but overall it looks good! It's a big dataset - and the image_tab information is still missing (and I think there'll be quite a lot of information there!). Also, a quick question. I noticed that the cell_suspensions were linked directly to an analysis_file, rather than to sequence_files. I assume the sequence_files were unavailable?
> yep, that's correct

**Comment by ami-day** [comment]
Thanks @Wkt8 , have made the updates suggested.

**Comment by ami-day** [comment]
Alegria and I are looking into an issue in ingest.

**Comment by ami-day** [comment]
syncing fastq to ingest prod.

**Comment by ami-day** [comment]
missing files: b2d53ada-0098-476b-bd00-debc030be73e

**Comment by ami-day** [comment]
Waiting to hear back from Sami about 9 of the image file names which have white space. Going to submit this project without those few files and update the project later when he gets back.

**Comment by ami-day** [comment]
graph validating

**Comment by ami-day** [comment]
exported.

**Comment by MightyAx** [comment]
Trying to export this project took down ingest.
A new operations ticket (and possibly new development) will be needed so that we can export this project to terra.

**Comment by Wkt8** [comment]
To be graph-validated and exported today

**Comment by MightyAx** [comment]
Pushed back to Graph Valid:
```
PUT https://api.ingest.archive.data.humancellatlas.org/submissionEnvelopes/626050ad6357205eb7bbc1e5/commitGraphValidEvent
kubectl rollout restart deployment ingest-state-tracking
```

**Comment by MightyAx** [comment]
I think this export is failing not because of size or memory issues (although those certainly weren't helping) but because of an Incomplete read error

THe logs disappeared before I could capture the details, but 4 exporter experiments errored for an incomplete read error, I managed to grab a partial stack track for one of the errors:

```
{"log":"During handling of the above exception, another exception occurred:\n","stream":"stdout","time":"2022-05-20T11:33:21.601972695Z"}
{"log":"\n","stream":"stdout","time":"2022-05-20T11:33:21.601967435Z"}
{"log":"http.client.IncompleteRead: IncompleteRead(7840 bytes read, 352 more expected)\n","stream":"stdout","time":"2022-05-20T11:33:21.601962744Z"}
{"log":" raise IncompleteRead(b''.join(s), amt)\n","stream":"stdout","time":"2022-05-20T11:33:21.601958314Z"}
{"log":" File \"/usr/local/lib/python3.7/http/client.py\", line 626, in _safe_read\n","stream":"stdout","time":"2022-05-20T11:33:21.601952834Z"}
{"log":" returned_chunk = self._fp._safe_read(self.chunk_left)\n","stream":"stdout","time":"2022-05-20T11:33:21.601949084Z"}
{"log":" File \"/usr/local/lib/python3.7/site-packages/urllib3/response.py\", line 723, in _handle_chunk\n","stream":"stdout","time":"2022-05-20T11:33:21.601944754Z"}
{"log":" chunk = self._handle_chunk(amt)\n","stream":"stdout","time":"2022-05-20T11:33:21.601940904Z"}
{"log":" File \"/usr/local/lib/python3.7/site-packages/urllib3/response.py\", line 770, in read_chunked\n","stream":"stdout","time":"2022-05-20T11:33:21.601936954Z"}
{"log":" yield\n","stream":"stdout","time":"2022-05-20T11:33:21.601933074Z"}
{"log":" File \"/usr/local/lib/python3.7/site-packages/urllib3/response.py\", line 441, in _error_catcher\n","stream":"stdout","time":"2022-05-20T11:33:21.601928404Z"}
{"log":"Traceback (most recent call last):\n","stream":"stdout","time":"2022-05-20T11:33:21.601923244Z"}
{"log":" 2022-05-20 11:33:21,576 - exporter.terra.terra_listener - ERROR in terra_listener.py:93 _experiment_message_handler(): ('Connection broken: IncompleteRead(7840 bytes read, 352 more expected)', IncompleteRead(7840 bytes read, 352 more expected))\n","stream":"stdout","time":"2022-05-20T11:33:21.601887494Z"}
{"log":" 2022-05-20 11:33:21,576 - exporter.terra.terra_listener - ERROR in terra_listener.py:92 _experiment_message_handler(): Failed to export experiment message with body: {\"exportJobId\":\"6287729242ceed65ffa2ce40\",\"documentId\":\"626055306357205eb7bbf268\",\"documentUuid\":\"66889875-8c93-48e7-80ff-60aecbee5e09\",\"callbackLink\":\"/processes/626055306357205eb7bbf268\",\"documentType\":\"process\",\"envelopeId\":\"626050ad6357205eb7bbc1e5\",\"envelopeUuid\":\"9c4eea83-8939-4232-9568-d31eea06718d\",\"projectId\":\"626050c46357205eb7bbc1e7\",\"projectUuid\":\"61515820-5bb8-45d0-8d12-f0850222ecf0\",\"index\":1837,\"total\":5174,\"context\":null}\n","stream":"stdout","time":"2022-05-20T11:33:21.583802259Z"}
```

**Comment by amnonkhen** [comment]
@ke4 to help with exporting again and tracking the progress

**Comment by amnonkhen** [comment]
The number of files in terra is 6388, the same as in the upload area, but the submission status remains "Exporting".

```
amnon@C02DW7FCMD6R ~ % aws s3 ls s3://org-hca-data-archive-upload-prod/9c4eea83-8939-4232-9568-d31eea06718d/ --profile embl-ebi | wc -l
    6388
amnon@C02DW7FCMD6R ~ % gsutil ls gs://broad-dsp-monster-hca-prod-ebi-storage/prod/61515820-5bb8-45d0-8d12-f0850222ecf0/data | wc -l
 
    6388
```

**Comment by amnonkhen** [comment]
Looking at the logs from 25/5, @ke4 @prabh-t and @amnonkhen noticed a restart of core immediately after initiating the export.
Another attempt to export at 26/5 around noon resulted in the same symptom of a series of restarts of core.

A possible theory is that something in the project is causing this. More updates to come.

**Comment by MightyAx** [comment]
Investigating the current export state of this before trying again:
## Comparing against the Submission Manifest
https://api.ingest.archive.data.humancellatlas.org/submissionEnvelopes/626050ad6357205eb7bbc1e5/submissionManifest

### Data
The data transfer operation has complete 🎉 
so the terra bucket /data folder should be left alone 👍 

### The Metadata
Using the following to list the metadata in terra
```
➜  gsutil ls -r gs://broad-dsp-monster-hca-prod-ebi-storage/prod/61515820-5bb8-45d0-8d12-f0850222ecf0/metadata | grep -v /: | cut -d/ -f7 | sed -r '/^\s*$/d' | uniq -c | sort -k2,2
```
#### Expected Biomaterials 2086
 168 cell_suspension
  32 donor_organism
1037 specimen_from_organism
 849 imaged_specimen
**Total: 2086** 👍 

#### Expected Processes 7237
18173 process files
But many are duplicated with newer datetime stamps in their file names from multiple exports:
**Total: 7224** after deduplication, still missing some 👎 

#### Expected Files 6397
  10 analysis_file
3946 image_file
2428 sequence_file
**Total: 6384**, still missing some 👎 

#### Expected Protocols 36
3 analysis_protocol
5 collection_protocol
4 dissociation_protocol
2 enrichment_protocol
7 imaging_preparation_protocol
4 imaging_protocol
4 library_preparation_protocol
5 sequencing_protocol
**Total: 34**, still missing some 👎 

### The Descriptors 6397
```
➜  gsutil ls -r gs://broad-dsp-monster-hca-prod-ebi-storage/prod/61515820-5bb8-45d0-8d12-f0850222ecf0/descriptors | grep -v /: | cut -d/ -f7 | sed -r '/^\s*$/d' | uniq -c | sort -k2,2
```
   9 analysis_file
3946 image_file
2428 sequence_file
**Total: 6383**, still missing some 👎 
Also 1 missing from the incomplete files 👎 

### The Links 76295
```
➜  gsutil ls gs://broad-dsp-monster-hca-prod-ebi-storage/prod/61515820-5bb8-45d0-8d12-f0850222ecf0/links | wc -l 
```
**16102** still missing some 👎

**Comment by MightyAx** [comment]
I'm going to:

- [ ] Scale the exporters to 0
- [ ] start an export job
- [ ] mark the job as `isDataTransferCcomplete: yes`
- [ ] Scale the exporters to 1
- [ ] Check that ingest does not crash
- [ ] slowly scale the Exporters back to 10

I'm hoping the export messages for the experiments that are in error can be investigated from the error queue.

**Comment by MightyAx** [comment]
At one point I had 20 exporter pods at once which made ingest hit it's maximum memory and crash,
thankfully this put the messages that failed on the error queue as expected.

**Comment by MightyAx** [comment]
Each of the export jobs finished and reported success, possibly because  let the first job run in isolation before adding more pods running in parallel.

Matching the submission manifest to the exported files on terra I can see we are missing:
- 2 processes, and
- 9 files

BUT! Since navigating the ingest UI for this submission I can also see that totals for processes and data files being off:
- 34 / 36 processes
- 6,388 / 6397 data files

And since the count of:
- `/data` files
- `/descriptors` files
- `/metadata/*_file` files
All match the count of 6388

**I believe the Submission manifest is currently incorrect.**

**Comment by MightyAx** [comment]
There is also currently the issue of the `/process` folder.
The (possibly incorrect) manifest references 7,237 processes,

the current process folder contains 23,347 files, but due to the date time stamp many of this files may be duplicates.

Of interesting not is that the most recent export only contained 5,174 assays

Current task: Calculate the real number of exported processes using a deduplication script

**Comment by MightyAx** [comment]
Missing URIs in file list: {'2c38512f-9d32-461d-8461-ea7d4731eb03': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055006357205eb7bbf00a', 'bf54cdb9-8048-4966-8602-eec27242c56d': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055036357205eb7bbf07d', '227f9d7c-78b2-4e80-bf73-70038c228515': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055156357205eb7bbf162', '753591f6-e0b0-4323-bafc-74e817ea04ef': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055246357205eb7bbf1e6', 'eace44cc-b27e-47df-adc1-7bbffdce5e17': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055536357205eb7bbf490', 'e8a89ca6-dd53-4b63-a7ce-168b3bc6b49a': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055606357205eb7bbf501', 'c1d86677-3aad-4f5b-a704-55776161b6ae': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055646357205eb7bbf582', '04591b40-6bf4-4838-9538-254ccbeb6784': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055726357205eb7bbf604', '35506675-7ac9-4b72-9d4e-0a15d64c501e': 'https://api.ingest.archive.data.humancellatlas.org/processes/626055806357205eb7bbf69b'}

**Comment by ofanobilbao** [comment]
@ami-day can you also, please, submit the import form, if all ok and you have not done it yet? Thanks!

**Comment by ami-day** [comment]
Yes but first I need to speak with @MightyAx about it and he is off sick today.

**Comment by ESapenaVentura** [comment]
@ami-day to take a look at the processes and why it states incorrect number of protocols/files

**Comment by gabsie** [comment]
@ami-day - please check the above message from @ESapenaVentura  - and hopefully you do not need Alexie for this check. Please confirm.

**Comment by ami-day** [comment]
> @ami-day - please check the above message from @ESapenaVentura - and hopefully you do not need Alexie for this check. Please confirm.

I have specific questions for Alexie about this, I'm going to discuss with him when he's back

**Comment by ami-day** [comment]
https://docs.google.com/spreadsheets/d/1DKVGFAwP3bcA4Atk9kW2LQe4rKahWW4B/edit#gid=2113381385

**Comment by MightyAx** [comment]
Notes from talking to Ami:
The 9 missing processes were for files that the author had added to the project, but were no longer required and not available for download. (Screencaps for presentations)

To make sure that the input biomaterials are correctly linked we should remove 9 processes from the project.

Any of the biomaterials that are linked to these processes should still then be lined to a remaining process.

Which will get detected by the unliked entity test as part of metadata validation (and graph validation if required)

If any specimens are unlinked they should be removed from terra.

**For future improvement, add a step in graph validation that checked that all processes have an output (either biomaterial or file)**

**Comment by MightyAx** [comment]
- given the following processes:

```

"processes/626055006357205eb7bbf00a"
"processes/626055036357205eb7bbf07d"
"processes/626055156357205eb7bbf162"
"processes/626055246357205eb7bbf1e6"
"processes/626055536357205eb7bbf490"
"processes/626055606357205eb7bbf501"
"processes/626055646357205eb7bbf582"
"processes/626055726357205eb7bbf604"
"processes/626055806357205eb7bbf69b"
```
- [ ] List all biomaterials linked to them
- [ ] Delete the processes
- [ ] Check the biomaterials are linked to a remaining object

**Comment by MightyAx** [comment]
Happy to say that all of the biomaterials that were linked to the now deleted processes were also linked to process that still remain.

That's (finally) everything done for this project!

**Comment by gabsie** [comment]
@MightyAx to just write an improvement ticket for graph validation  - to check the process is linked not only to a sequence file but to an image file as well

**Comment by MightyAx** [comment]
@ami-day to fill input form to add this project to release 20.

**Comment by ami-day** [comment]
The project seems to be in "metadata valid" state https://contribute.data.humancellatlas.org/projects/detail?uuid=61515820-5bb8-45d0-8d12-f0850222ecf0&tab=upload as oppose to "exported"

**Comment by MightyAx** [comment]
I imagine graph validation might fail:
- that okay
- we can run a "manual" graph validation again
- and then move the project manually to exported

**Comment by MightyAx** [comment]
Checking the logs the project failed graph validation 10am yesterday and took down ingest. I'll try running a manual graph validation on the ec2 now.

**Comment by ESapenaVentura** [comment]
@ami-day @MightyAx We need to re-export the project. I have noted that there are double pipes in the `computational_method` field of the analysis protocols, so we'll need to fix that and re-export

**Comment by ami-day** [comment]
> Checking the logs the project failed graph validation 10am yesterday and took down ingest. I'll try running a manual graph validation on the ec2 now.

Thank you!

**Comment by idazucchi** [comment]
@MightyAx will set this to metadata valid so that Ami can remove the double pipe

**Comment by MightyAx** [comment]
I've set this back to metadata valid and restarted state tracking
Please do not progress to graph validation or exporting until the current exporting issues have been resolved

**Comment by gabsie** [comment]
Whenever Ami does updates, she contacts Alexie for a manual graph validation.

**Comment by ami-day** [comment]
Updates done, and have messaged Alexie to let him know.

**Comment by MightyAx** [comment]
Using the same method as in ticket #757,

1. I've emptied the neo4j database
2. initialised graph validation environment
3. begun hydration (populating the data).

Last time the hydration toon 9 hours so tomorrow we can check in and hopefully

4. start graph validation
5. mark submission as graph valid

**Comment by MightyAx** [comment]
This project is graph valid:
```
(ingest-graph-validator) ubuntu@ip-172-31-71-222:~/ingest-graph-validator$ ingest-graph-validator hydrate ingest 9c4eea83-8939-4232-9568-d31eea06718d
22-08-08 11:00:45 [ingest_graph_validator.ingest_graph_validator] - INFO: starting neo4j docker instance
22-08-08 11:00:45 [ingest_graph_validator.ingest_graph_validator] - INFO: attached to backend container [neo4j-server]
22-08-08 11:00:45 [ingest_graph_validator.ingest_graph_validator] - INFO: Connecting to neo4j...
22-08-08 11:00:45 [ingest_graph_validator.hydrators.hydrator] - INFO: Started ingest hydrator for for submission [9c4eea83-8939-4232-9568-d31eea06718d]
22-08-08 11:00:45 [ingest_graph_validator.hydrators.hydrator] - INFO: Found project for submission 61515820-5bb8-45d0-8d12-f0850222ecf0
22-08-08 11:00:45 [ingest_graph_validator.hydrators.hydrator] - INFO: Found submission for project with uuid 9c4eea83-8939-4232-9568-d31eea06718d
22-08-08 12:32:08 [ingest_graph_validator.hydrators.hydrator] - INFO: imported 15737 nodes
22-08-08 12:32:08 [ingest_graph_validator.utils] - INFO: [get_nodes] took [581] ms
22-08-08 13:11:17 [ingest_graph_validator.hydrators.hydrator] - INFO: imported 113372 edges
22-08-08 13:11:17 [ingest_graph_validator.utils] - INFO: [get_edges] took [2349133] ms
22-08-08 20:01:22 [ingest_graph_validator.hydrators.hydrator] - INFO: hydration finished

(ingest-graph-validator) ubuntu@ip-172-31-71-222:~/ingest-graph-validator$ ingest-graph-validator action test graph_test_set
22-08-10 08:33:30 [ingest_graph_validator.ingest_graph_validator] - INFO: starting neo4j docker instance
22-08-10 08:33:30 [ingest_graph_validator.ingest_graph_validator] - INFO: attached to backend container [neo4j-server]
22-08-10 08:33:30 [ingest_graph_validator.ingest_graph_validator] - INFO: Connecting to neo4j...
22-08-10 08:33:30 [ingest_graph_validator.actions.test_action] - INFO: loading tests
22-08-10 08:33:30 [ingest_graph_validator.actions.test_action] - INFO: loaded [16] test queries
22-08-10 08:33:30 [ingest_graph_validator.actions.test_action] - INFO: running tests
22-08-10 08:33:32 [ingest_graph_validator.actions.test_action] - INFO: All tests finished
```

**Comment by MightyAx** [comment]
Project Set to GraphValid:
```
curl -L -X PUT "https://api.ingest.archive.data.humancellatlas.org/submissionEnvelopes/626050ad6357205eb7bbc1e5/commitGraphValidEvent" -H "Authorization: Bearer <snip>"
```

**Comment by MightyAx** [comment]
I've exported the project, I can see that it is marked as exported in ingest, and new versions of the analysis_protocols have been uploaded to terra.

**Comment by ofanobilbao** [comment]
Yesterday I identified that the submission was back to "Metadata valid", and that we had 2 extra duplicates of the project in Ingest. One of them with a "Metadata invalid" submission and one that looked as the originally triaged copy, as all the information on the Project tab was filled out and it was the only one with the tick to be displayed in the Catalogue. After a chat after stand-up today, the status is as follows:

- I have copied across all the missing fields to the copy with the valid submission, and I have ticked the Catalogue box.
- The submission status has been changed to exported by Alexie
- Ami has submitted the import form for Release 21
- I have deleted the duplicate that did not have any submissions (I need to inform Will)
- I have created a ticket in the Ops board to delete the other Project duplicate as I have not been able to delete the submission that was "Metadata invalid" and therefore have not been able to delete the Project

**Comment by ami-day** [comment]
looks ok

**Comment by ofanobilbao** [comment]
Was affected by double pipes and removed from DCP Portal. Has been fixed and exported again now for R28. @ESapenaVentura will fill the import form as agreed with Mary Wons that it will go into R28

**Comment by ofanobilbao** [comment]
Issues were flagged again on indexing of R28. The project was dropped. Fixing is being tracked in an Ops ticket, https://app.zenhub.com/workspaces/operations-5fa2d8f2df78bb000f7fb2b5/issues/gh/ebi-ait/hca-ebi-wrangler-central/1127

**Comment by ofanobilbao** [comment]
Enrique just confirmed he managed to export this dataset again for R29

**Comment by ofanobilbao** [comment]
Moving to close as Browser team did not flag any double pipes or metadata issues
